src/VirtoCommerce.Platform.Core/JsonConverters/PolymorphJsonConverter.cs

```diff
@@ -1,49 +1,76 @@
-            object result;
-            var obj = JObject.Load(reader);
-            if (typeof(PermissionScope).IsAssignableFrom(objectType))
+
+            RegisterType(type, obj =>
             {
-                var scopeType = objectType.Name;
-                var pt = obj["type"] ?? obj["Type"];
+                var typeName = type.Name;
+                var pt = obj[descriminator] ?? obj[descriminator.FirstCharToUpper()];
                 if (pt != null)
                 {
-                    scopeType = pt.Value<string>();
+                    typeName = pt.Value<string>();
                 }
-                result = AbstractTypeFactory<PermissionScope>.TryCreateInstance(scopeType);
+                var tryCreateInstance = typeof(AbstractTypeFactory<>).MakeGenericType(type).GetMethods().FirstOrDefault(x => x.Name.Equals("TryCreateInstance") && x.GetParameters().Length == 1);
+                var result = tryCreateInstance?.Invoke(null, new[] { typeName });
                 if (result == null)
                 {
-                    throw new NotSupportedException("Unknown scopeType: " + scopeType);
+                    throw new NotSupportedException("Unknown scopeType: " + typeName);
                 }
-            }
-            else
+                return result;
+            });
+
+        }
+        public static void RegisterType(Type type, Func<JObject, object> factory)
+        {
+            _canConvertCache[type] = true;
+            _convertFactories[type] = factory;
+        }
+
+        public override bool CanConvert(Type objectType)
+        {
+            if (objectType.IsPrimitive || objectType.Equals(typeof(string)) || objectType.IsArray)
             {
-                var tryCreateInstance = typeof(AbstractTypeFactory<>).MakeGenericType(objectType).GetMethods().FirstOrDefault(x => x.Name.EqualsInvariant("TryCreateInstance") && x.GetParameters().Length == 0);
-                result = tryCreateInstance?.Invoke(null, null);
+                return false;
             }
+            var result = _canConvertCache.GetOrAdd(objectType, _ =>
+            {
+                return (bool)typeof(AbstractTypeFactory<>).MakeGenericType(objectType).GetProperty("HasOverrides").GetValue(null, null);
+            });
+            return result;
+        }
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            object result;
+            var obj = JObject.Load(reader);
+
+            object Factory(JObject obj2)
+            {
+                //TODO: Optmimize reflection
+                var tryCreateInstance = typeof(AbstractTypeFactory<>).MakeGenericType(objectType).GetMethods().FirstOrDefault(x => x.Name.Equals("TryCreateInstance") && x.GetParameters().Length == 0);
+                return tryCreateInstance?.Invoke(null, null);
+            };
+            var factory = _convertFactories.GetOrAdd(objectType, _ => Factory);
```

Comment:
no need for named function, put the implementation as parameter. Same as for "_canConvertCache.GetOrAdd(objectType, _ =>"